### ISSUE TITLE
Groups missing from claims

### ISSUE THREAD

**Issue by robalexclark** [okta/okta-aspnet#143]
I'm using the following startup.cs config for okta:

      ```
      services.AddAuthentication(options =>
            {
                options.DefaultAuthenticateScheme = CookieAuthenticationDefaults.AuthenticationScheme;
                options.DefaultSignInScheme = CookieAuthenticationDefaults.AuthenticationScheme;
                options.DefaultChallengeScheme = OktaDefaults.MvcAuthenticationScheme;
            })
            .AddCookie()
            .AddOktaMvc(new OktaMvcOptions
            {
                // Replace these values with your Okta configuration
                OktaDomain = Configuration.GetValue<string>("Okta:OktaDomain"),
                ClientId = Configuration.GetValue<string>("Okta:ClientId"),
                ClientSecret = Configuration.GetValue<string>("Okta:ClientSecret"),
                Scope = new List<string> { "openid", "profile", "email", "groups" },
                GetClaimsFromUserInfoEndpoint = true                
            });
``

However whenever I look at the ClaimsPrincipal for the groups claim no such claim exists:



``
  AuthenticationState authenticationState = await authenticationStateProvider.GetAuthenticationStateAsync();
            var claimsList = authenticationState.User.Claims;

            var userIdentity = (ClaimsIdentity)authenticationState.User.Identity;
            var claims = userIdentity.Claims;
            var roleClaimType = userIdentity.RoleClaimType;
            var roles = claims.Where(c => c.Type == ClaimTypes.Role).ToList();
``

Can anyone advise why?

**Comment by bryanapellanes-okta** [comment]
@robalexclark ,
Thanks for your inquiry!  I will review this more closely to provide you an answer.  Stay tuned!

**Comment by bryanapellanes-okta** [comment]
@robalexclark,
Further investigation has revealed that `groups` scope is not configured by default.  You can learn more about working with Okta oidc scopes from the following links:

- [Scopes and supported endpoints](https://developer.okta.com/docs/guides/implement-oauth-for-okta/scopes/)
- [Create Scopes](https://developer.okta.com/docs/guides/customize-authz-server/create-scopes/)
- [Define allowed scopes](https://developer.okta.com/docs/guides/implement-oauth-for-okta/define-allowed-scopes/)

Hope this helps.

**Comment by robalexclark** [comment]
We do have a groups scope configured and other (non .net core) applications are getting the groups in their list of claims.

**Comment by bryanapellanes-okta** [comment]
> We do have a groups scope configured and other (non .net core) applications are getting the groups in their list of claims.

Can you provide a working code example showing the retrieval of groups in the list of claims?  I can use that to determine why it isn't working with .net core.

**Comment by robalexclark** [comment]
Unfortunately I can't provide a working example due to my companies regulations. However they have opened a support ticket with okta to try and resolve.

One question - would you expect the groups to be included in the list of claims automatically retrieved on successful login? I have seen examples on okta support where a separate api call is required to retrieve them, but I didn't know whether this was necessary with the okra-aspnet package.

**Comment by bryanapellanes-okta** [comment]
> One question - would you expect the groups to be included in the list of claims automatically retrieved on successful login? I have seen examples on okta support where a separate api call is required to retrieve them, but I didn't know whether this was necessary with the okra-aspnet package.

Generally the answer to your question is yes.  However, a difference has been seen in .Net Core due to a change in the way that tokens are processed by the underlying framework and related libraries resulting in the behavior you're seeing.  In this case, to retrieve the current user's groups you can reference the [Okta.Sdk](https://www.nuget.org/packages/Okta.Sdk/) and make a separate call using code similar to the following, making sure to inject `_configuration` as an instance of `IConfiguration` in your controller constructor:
```
        [Authorize]
        public async Task<IActionResult> Groups()
        {
            OktaClientConfiguration config = new OktaClientConfiguration
            {
                OktaDomain = _configuration.GetValue<string>("Okta:OktaDomain"),
                Token = _configuration.GetValue<string>("Okta:Token"),
            };
            string userId = HttpContext.User.Claims.FirstOrDefault(c => (bool)c.Type?.Equals("sub")).Value;
            List<IGroup> groups = await new OktaClient(config).Users.ListUserGroups(userId).ToListAsync();
            return View(groups);
        }
```
> ⚠️  Note &mdash; the OktaClientConfiguration.Token value expects an api token from your Okta developer dashboard for your application

**Comment by robalexclark** [comment]
We have resolved this now - the reason was that the application was setup to send an access token rather than an openid token. Once this was changed the claims started coming through. 

One tweak I had to make was that the Group claims sent thru were not marked as being ClaimType.Role so the IsInRole method would not work, I couldn't find a way to change a claim type without first manually removing and then recreating the claim. In the end I created a Policy where a user had to have a group claim rather than using roles directly.
